File: java/08_072.java
import java.util.*;

class Customer {
   String id;
   int partySize;
   int vipStars;
   boolean hasReservation;
   int arrivalTime;
   
   public Customer(String id, int partySize, int vipStars, 
         boolean hasReservation, int arrivalTime) {
      this.id = id;
      this.partySize = partySize;
      this.vipStars = vipStars;
      this.hasReservation = hasReservation;
      this.arrivalTime = arrivalTime;
   }
   
   public int getPriority(int currentTime) {
      // Calculate priority based on VIP status, reservation, wait time
      int waitingTime = currentTime - arrivalTime;
      return (vipStars * 10) + (hasReservation ? 20 : 0) + 
            waitingTime;
   }
}

class Result {
   String customerId;
   int seatedTime;
   
   public Result(String customerId, int seatedTime) {
      this.customerId = customerId;
      this.seatedTime = seatedTime;
   }
}

public class RestaurantPrioritySeating {
   public static List<Result> manageSeating(List<Integer> tables, 
         List<Object[]> events) {
      // Sort events chronologically
      events.sort(Comparator.comparing(e -> (Integer)e[1]));
      
      // Priority queue for waitlist (max heap)
      PriorityQueue<Customer> waitlist = new PriorityQueue<>(
            (c1, c2) -> Integer.compare(
                  c2.getPriority(0), c1.getPriority(0)));
      
      List<Result> results = new ArrayList<>();
      
      for (Object[] event : events) {
         String eventType = (String)event[0];
         int currentTime = (Integer)event[1];
         
         // Update waitlist priorities
         PriorityQueue<Customer> updatedWaitlist = 
               new PriorityQueue<>((c1, c2) -> 
                     Integer.compare(
                        c2.getPriority(currentTime), 
                        c1.getPriority(currentTime)));
         updatedWaitlist.addAll(waitlist);
         waitlist = updatedWaitlist;
         
         if (eventType.equals("ARRIVE")) {
            // Process customer arrival
            String customerId = (String)event[2];
            int partySize = (Integer)event[3];
            int vipStars = (Integer)event[4];
            boolean hasReservation = (Boolean)event[5];
            
            Customer newCustomer = new Customer(
                  customerId, partySize, vipStars, 
                  hasReservation, currentTime);
            waitlist.add(newCustomer);
         } else if (eventType.equals("TABLE_READY")) {
            // Seat customers at available table
            int tableSize = (Integer)event[2];
            List<Customer> temp = new ArrayList<>();
            Customer seated = null;
            
            while (!waitlist.isEmpty() && seated == null) {
               Customer customer = waitlist.poll();
               
               if (customer.partySize <= tableSize) {
                  seated = customer;
                  results.add(new Result(customer.id, 
                        currentTime));
               } else {
                  temp.add(customer);
               }
            }
            
            // Restore unseated customers
            waitlist.addAll(temp);
         }
      }
      
      return results;
   }
}

File: python/08_072.py
import heapq

def restaurant_priority_seating(tables, events):
   # Max heap in Python (implemented as min heap with negatives)
   waitlist = []  # (-priority_score, customer_id, party_size,
                  #  arrival_time, vip_stars, has_reservation)
   results = []
   
   # Process events in chronological order
   for event in events:
      event_type = event[0]
      current_time = event[1]
      
      # Update all customers' priorities based on waiting time
      updated_waitlist = []
      while waitlist:
         neg_priority, customer_id, party_size, arrival_time, \
         vip_stars, has_reservation = heapq.heappop(waitlist)
         waiting_time = current_time - arrival_time
         new_priority = (vip_stars * 10) + \
                        (20 if has_reservation else 0) + waiting_time
         heapq.heappush(updated_waitlist, (
            -new_priority, customer_id, party_size,
            arrival_time, vip_stars, has_reservation
         ))
      waitlist = updated_waitlist
      
      if event_type == "ARRIVE":
         customer_id = event[2]
         party_size = event[3]
         vip_stars = event[4]
         has_reservation = event[5]
         
         # Calculate initial priority
         initial_priority = (vip_stars * 10) + \
                           (20 if has_reservation else 0)
         
         # Add to waitlist
         heapq.heappush(waitlist, (
            -initial_priority, customer_id, party_size,
            current_time, vip_stars, has_reservation
         ))
         
      elif event_type == "TABLE_READY":
         table_size = event[2]
         
         # Find highest priority customer that fits at this table
         temp_waitlist = []
         seated = False
         
         while waitlist and not seated:
            neg_priority, customer_id, party_size, arrival_time, \
            vip_stars, has_reservation = heapq.heappop(waitlist)
            
            if party_size <= table_size:
               # Seat this customer
               results.append([customer_id, current_time])
               seated = True
            else:
               # Put back in temporary waitlist
               heapq.heappush(temp_waitlist, (
                  neg_priority, customer_id, party_size,
                  arrival_time, vip_stars, has_reservation
               ))
         
         # Restore any customers we examined but didn't seat
         while temp_waitlist:
            heapq.heappush(waitlist, heapq.heappop(temp_waitlist))
   
   return results